p2p/log.go

```diff
@@ -9,100 +14,131 @@ const (
+	if err != nil {
+		log.Debug(err.Error())
+		return nil, err
+	}
+
+	// iterate through the peer ids
+	for _, pid := range pids {
+
+		if err = n.SendMessage(req, replies, pid); err != nil {
+			log.Debugf("%s err: %s", pid, err.Error())
+			continue
+		}
+
+		res := <-replies
+		dlrb := DatasetLogResponseBody{}
+		if err = json.Unmarshal(res.Body, &dlrb); err == nil {
+			if dlrb.Err != nil {
+				log.Debugf("%s err: %s", pid, err.Error())
+				continue
+			}
+			if len(dlrb.History) != 0 {
+				return dlrb.History, nil
+			}
+		}
+	}
+	return nil, fmt.Errorf("unable to locate dataset log for %s", ref)
 }
 
-func (n *QriNode) datasetsHistoryHandler(ws *WrappedStream, msg Message) (hangup bool) {
-	// data, err := json.Marshal(msg.Payload)
-	// if err != nil {
-	// 	log.Debug(err.Error())
-	// }
-
-	return false
-
-	// ref := repo.DatasetRef{}
-	// if err = json.Unmarshal(data, &ref); err != nil {
-	// 	log.Infof(err.Error())
-	// 	return &Message{
-	// 		Type:    MtDatasetLog,
-	// 		Phase:   MpError,
-	// 		Payload: err,
-	// 	}
-	// }
-
-	// ref, err = n.Repo.GetRef(ref)
-	// if err != nil {
-	// 	return &Message{
-	// 		Type:    MtDatasetLog,
-	// 		Phase:   MpError,
-	// 		Payload: err,
-	// 	}
-	// }
-	// // TODO: probably shouldn't write over ref.Path if ref.Path is set, but
-	// // until we make the changes to the way we use hashes to make them
-	// // more consistent, this feels safer.
-	// // ref.Path = path.String()
-
-	// log := []repo.DatasetRef{}
-	// limit := 50
-
-	// for {
-	// 	ref.Dataset, err = n.Repo.GetDataset(datastore.NewKey(ref.Path))
-	// 	if err != nil {
-	// 		return &Message{
-	// 			Type:    MtDatasetLog,
-	// 			Phase:   MpError,
-	// 			Payload: err,
-	// 		}
-	// 	}
-	// 	log = append(log, ref)
-
-	// 	limit--
-	// 	if limit == 0 || ref.Dataset.PreviousPath == "" {
-	// 		break
-	// 	}
-
-	// 	ref, err = repo.ParseDatasetRef(ref.Dataset.PreviousPath)
-
-	// 	if err != nil {
-	// 		return &Message{
-	// 			Type:    MtDatasetLog,
-	// 			Phase:   MpError,
-	// 			Payload: err,
-	// 		}
-	// 	}
-	// }
-	// return &Message{
-	// 	Type:    MtDatasetLog,
-	// 	Phase:   MpResponse,
-	// 	Payload: &log,
-	// }
+func (n *QriNode) handleDatasetLog(ws *WrappedStream, req Message) (hangup bool) {
+	hangup = true
+
+	switch req.Header("phase") {
+	case "request":
+
+		resBody := DatasetLogRequestBody{}
+		if err := json.Unmarshal(req.Body, &resBody); err != nil {
+			log.Debug(err.Error())
+			return
+		}
+
+		ref := resBody.Ref
+		limit := resBody.Limit
+		offset := resBody.Offset
+
+		local := true
+
+		err := n.ResolveDatasetRef(&ref)
+		if err != nil {
+			local = false
+		}
+
+		reqBody := DatasetLogResponseBody{}
+		history := []repo.DatasetRef{}
+
+		if local {
+			for {
+				dataset, err := dsfs.LoadDataset(n.Repo.Store(), datastore.NewKey(ref.Path))
+				if err != nil {
+					reqBody.Err = err
+					break
+				}
+				ref.Dataset = dataset.Encode()
+
+				offset--
+				if offset > 0 {
+					continue
```

Comment:
Is this supposed to be setting ref.Path to ref.Dataset.PreviousPath? Otherwise a non-zero offset will just keep loading the same ref.Path over and over.